File: php/src/Advertisements/Advertisement/Application/Command/EnableAdvertisement/EnableAdvertisementUseCase.php
<?php
declare(strict_types=1);

namespace Demo\App\Advertisements\Advertisement\Application\Command\EnableAdvertisement;

use Demo\App\Advertisements\Advertisement\Domain\AdvertisementRepository;
use Demo\App\Advertisements\Advertisement\Domain\Exceptions\AdvertisementNotFoundException;
use Demo\App\Advertisements\Advertisement\Domain\ValueObjects\AdvertisementId;
use Demo\App\Advertisements\Shared\Domain\ValueObjects\UserId;
use Demo\App\Advertisements\User\Domain\Exceptions\AdminWithIncorrectCivicCenterException;
use Demo\App\Advertisements\User\Domain\Exceptions\MemberDoesNotExistsException;
use Demo\App\Advertisements\User\Domain\Exceptions\UserNotFoundException;
use Demo\App\Advertisements\User\Domain\UserRepository;
use Exception;

final class EnableAdvertisementUseCase
{
    public function __construct(
        private AdvertisementRepository $advertisementRepository,
        private UserRepository $userRepository
    ) {}

    /**
     * @throws Exception
     */
    public function execute(EnableAdvertisementCommand $command): void
    {
        //TODO: Complex behaviour
        //TODO: Find possible bug
        $adminUser = $this->userRepository->findAdminById(new UserId($command->securityUserId));
        if (!$adminUser) {
            throw UserNotFoundException::asAdmin();
        }

        $advertisement = $this->advertisementRepository->findById(new AdvertisementId($command->advertisementId));

        if (!$advertisement) {
            throw AdvertisementNotFoundException::withId($command->advertisementId);
        }

        $member = $this->userRepository->findMemberById($advertisement->memberId());

        if (null === $member) {
            throw MemberDoesNotExistsException::build();
        }

        if (!$adminUser->civicCenterId()->equals($member->civicCenterId())) {
            throw AdminWithIncorrectCivicCenterException::differentCivicCenterFromMember();
        }

        $activeAdvertisements = $this->advertisementRepository->activeAdvertisementsByMember($member);

        if ($activeAdvertisements->value() >= 3) {
            throw new Exception('Member has reached the maximum number of active advertisements');
        }

        $advertisement->enable();

        $this->advertisementRepository->save($advertisement);
    }
}

File: kotlin/src/main/kotlin/advertisements/advertisement/application/command/enableadvertisement/EnableAdvertisementUseCase.kt
package advertisements.advertisement.application.command.enableadvertisement

import advertisements.advertisement.domain.exceptions.AdvertisementNotFoundException
import advertisements.advertisement.domain.AdvertisementRepository
import advertisements.advertisement.domain.services.AdvertisementSecurityService
import advertisements.advertisement.domain.value_object.AdvertisementId
import advertisements.shared.value_object.UserId
import advertisements.user.domain.UserRepository
import advertisements.user.domain.exceptions.UserNotFoundException

class EnableAdvertisementUseCase(
    private val advertisementRepository: AdvertisementRepository,
    private val userRepository: UserRepository,
    private val advertisementSecurityService: AdvertisementSecurityService
) {
    fun execute(enableAdvertisementCommand: EnableAdvertisementCommand) {
        val advertisementId = AdvertisementId(enableAdvertisementCommand.advertisementId)
        val advertisement = advertisementRepository.findById(advertisementId)

        if (null === advertisement) {
            throw AdvertisementNotFoundException.withId(advertisementId.value())
        }

        advertisementSecurityService.verifyAdminUserCanManageAdvertisement(
            UserId(enableAdvertisementCommand.securityUserId),
            advertisement
        )

        val memberUser = userRepository.findMemberById(advertisement.memberId)
        if (memberUser == null) {
            throw UserNotFoundException.asMember()
        }

        val activeAdvertisements = advertisementRepository.activeAdvertisementsByMember(memberUser)

        if (activeAdvertisements.value() >= 3) {
            throw Exception("Member has reached the maximum number of active advertisements")
        }

        advertisement.enable()

        advertisementRepository.save(advertisement)
    }
}
